learning/tour-of-beam/frontend/lib/pages/tour/state.dart

```diff
@@ -0,0 +1,172 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import 'package:app_state/app_state.dart';
+import 'package:flutter/widgets.dart';
+import 'package:get_it/get_it.dart';
+import 'package:playground_components/playground_components.dart';
+
+import '../../cache/unit_content.dart';
+import '../../models/unit.dart';
+import '../../models/unit_content.dart';
+import 'controllers/content_tree.dart';
+import 'path.dart';
+
+// Copied from Playground's config.g.dart
+// TODO(alexeyinkin): Read some config when backend's URL is emitted from
+//  an infrastructure workflow.
+
+const String kApiClientURL = 'https://router.34.160.22.119.nip.io';
+const String kApiJavaClientURL =
+    'https://backend-java-beta-dot-apache-beam-testing.appspot.com';
+const String kApiGoClientURL =
+    'https://backend-go-beta-dot-apache-beam-testing.appspot.com';
+const String kApiPythonClientURL =
+    'https://backend-python-beta-dot-apache-beam-testing.appspot.com';
+const String kApiScioClientURL =
+    'https://backend-scio-beta-dot-apache-beam-testing.appspot.com';
+
+class TourNotifier extends ChangeNotifier with PageStateMixin<void> {
+  final ContentTreeController contentTreeController;
+  final PlaygroundController playgroundController;
+  final _unitContentCache = GetIt.instance.get<UnitContentCache>();
+  UnitContentModel? _currentUnitContent;
+
+  TourNotifier({
+    required String initialSdkId,
+    List<String> initialTreeIds = const [],
+  })  : contentTreeController = ContentTreeController(
+          initialSdkId: initialSdkId,
+          initialTreeIds: initialTreeIds,
+        ),
+        playgroundController = _createPlaygroundController(initialSdkId) {
+    contentTreeController.addListener(_onContentTreeControllerChanged);
+    _unitContentCache.addListener(_onChanged);
+  }
+
+  @override
+  PagePath get path => TourPath(
+        sdkId: contentTreeController.sdkId,
+        treeIds: contentTreeController.treeIds,
+      );
+
+  void _onContentTreeControllerChanged() {
+    _onChanged();
+  }
+
+  void _onChanged() {
+    final currentNode = contentTreeController.currentNode;
+    if (currentNode is UnitModel) {
+      final content = _unitContentCache.getUnitContent(
+        contentTreeController.sdkId,
+        currentNode.id,
+      );
+
+      _setCurrentUnitContent(content);
+    } else {
+      _emptyPlayground();
+    }
+
+    notifyListeners();
+  }
+
+  UnitContentModel? get currentUnitContent => _currentUnitContent;
```

Comment:
Getters should not be defined between methods in the meddle of the class. Move it to defining another class fields please

Reply:
It's fine here above a conceptually similar setter which is a method. It should only be moved with that setter. But then the private setter will be defined above its usage which is confusing.